### PR TITLE
Re-commit all commits of "escaping single quote format of shell command

### PR DESCRIPTION
**Pull request by liming01**
This PR is the latest update for below PRs which have been reverted for CI failure:
https://github.com/greenplum-db/gpdb/pull/5974
https://github.com/greenplum-db/gpdb/pull/6058
https://github.com/greenplum-db/gpdb/pull/5927

Notes in testcase about backslash escaping:
    - Need to add ESCAPE 'OFF' to EXTERNAL WEB TABLE
    - Need to add ESCAPE 'OFF' to COPY ... PROGRAM
    - Need to add ESCAPE '&' for LIKE predicate
    - Command echo will behaves differently on different platforms, force to use bash shell with -E option.
    - For shell 'env' output, don't seperate to 2 columns because CI env has funny char
      in variable value. e.g.
        "LS_OPTIONS=-N --color=tty -T 0" and
        "LESSOPEN=||/usr/bin/lesspipe.sh %s"

### CODE REVIEW

File: src/backend/access/external/fileam.c [hunk 1]
```diff
@@ -34,6 +34,7 @@
 #include "postgres.h"
 
 #include <fstream/gfile.h>
+#include <tcop/tcopprot.h>
```
Comment (hlinnaka) [inline src/backend/access/external/fileam.c]: Why is this needed?

File: src/backend/access/external/url_execute.c [hunk 2]
```diff
@@ -181,8 +181,10 @@ make_export(char *name, const char *value, StringInfo buf)
 
 	for ( ; 0 != (ch = *value); value++)
 	{
-		if (ch == '\'' || ch == '\\')
-			appendStringInfoChar(buf, '\\');
+		if(ch == '\'')
```
Comment (hlinnaka) [inline src/backend/access/external/url_execute.c]: IMHO it would still be nice to have a comment here, to explain the quoting works, like I had in commit 7a9de421c8. Given that we were both confused in different ways, on how the quoting should work, and we had to revert stuff, this stuff really is not obvious. It needs a longish comment to explain how the quoting works, and it would be good to mention the things we found out during the debugging and testing these cases, what we found out about "echo" quoting for example. So something like what I had in commit 7a9de421c8, but even more than that.

File: src/test/regress/input/external_table.source [hunk 3]
```diff
@@ -110,6 +111,12 @@ drop external table check_ps;
 drop external table check_env;
 
 -- end_ignore
+CREATE EXTERNAL WEB TABLE table_env (val TEXT)
```
Comment (hlinnaka) [inline src/test/regress/input/external_table.source]: Please add a comment to explain what this test is for.

File: src/backend/access/external/fileam.c (hunk 1 above)
Comment (liming01) [reply src/backend/access/external/fileam.c]: Yes, the variable "debug_query_string" defined in this file.

File: src/backend/access/external/url_execute.c (hunk 2 above)
Comment (liming01) [reply src/backend/access/external/url_execute.c]: Updated in the new commit.

**Comment by liming01** [comment]
Hi @hlinnaka ,

As your requests, I split it into several commits, which is basing on previous commit with refined enhancement/fix. Please help me to review it again.

File: src/test/regress/output/gpcopy.source [hunk 4]
```diff
@@ -1454,3 +1454,25 @@ drop type broken_int4 cascade; -- drops the I/O functions, too.
...
+-- as environment variables to the program. Test those.
+--
+-- Perform these tests in a funnily named database, to test
+-- escaping
+set client_min_messages='warning';
+DROP DATABASE IF EXISTS "funny copy""db'with\\quotes";
+reset client_min_messages;
+CREATE DATABASE "funny copy""db'with\\quotes";
+\c "funny copy""db'with\\quotes"
+COPY (SELECT 'data1') TO PROGRAM 'cat > /tmp/gpcopyenvtest; /bin/bash -c ''echo -E database in COPY TO: $GP_DATABASE >> /tmp/gpcopyenvtest '' ' ESCAPE 'OFF';
```
Comment (hlinnaka) [inline src/test/regress/output/gpcopy.source]: The https://wiki.ubuntu.com/DashAsBinSh#echo page recommends using printf instead. I think that's a reasonable suggestion. So `printf "database in COPY TO: %s\n" $GP_DATABASE`

File: src/test/regress/output/external_table.source [hunk 5]
```diff
@@ -81,6 +81,7 @@ select * from check_env;
  GP_HADOOP_CONN_VERSION=CE_1.0.0.0
  GP_MASTER_HOST=127.0.0.1
  GP_MASTER_PORT=5432
+ GP_QUERY_STRING=select * from check_env;
```
Comment (hlinnaka) [inline src/test/regress/output/external_table.source]: This is in a start/end_ignore block, so we're not actually having any test coverage to check that GP_QUERY_STRING really is set in the environment. I'm actually kind of OK with that, this is such a simple addition, but might nevertheless be good to have a test for it. It should be pretty straightforward to test, too; I think you had a test for it in earlier versions of this patch.

File: src/test/regress/output/external_table.source [hunk 6]
```diff
@@ -128,6 +128,37 @@ select * from check_env;
 (55 rows)
 
 -- end_ignore
+CREATE EXTERNAL WEB TABLE table_env (val TEXT)
+  EXECUTE E'env' ON SEGMENT 0
+  FORMAT 'TEXT' (ESCAPE 'OFF');
+SELECT * FROM table_env WHERE val LIKE 'GP_QUERY%' ORDER BY val ASC;
```
Comment (hlinnaka) [inline src/test/regress/output/external_table.source]: This test belongs in the other commit, in the one that adds the GP_QUERY_STRING variable

Comment (liming01) [reply src/test/regress/output/external_table.source]: Yes, but it need the fix of this commit.

File: src/test/regress/output/gpcopy.source (hunk 4 above)
Comment (liming01) [reply src/test/regress/output/gpcopy.source]: Our test is not only just for testing, but also for helping people who encounter problems. So I would like to use echo, because user is more likely to use echo than printf.

File: src/test/regress/output/external_table.source (hunk 5 above)
Comment (liming01) [reply src/test/regress/output/external_table.source]: I don't know why this block is ignored in original test, maybe the env setting are totally different on different CI. So this test is useless at all. Do we need to remove it?

**Comment by hlinnaka** [comment]
It looks correct to me now. But given the struggles we both had with this, I feel like we should get a third opinion. Also because I've been so involved in this, I'm not sure I would notice anymore, if the comments are not adequate to explain things to someone who's not already familiar with this.

@danielgustafsson, would you have time to review this? Or someone else?

File: src/backend/commands/copy.c [hunk 7]
```diff
@@ -7355,6 +7355,7 @@ open_program_pipes(char *command, bool forwrite)
 	/* Restore the SIGPIPE handler */
 	pqsignal(SIGPIPE, save_SIGPIPE);
 
+	elog(DEBUG5, "Command: %s", program_pipes->shexec);
```
Comment (danielgustafsson) [inline src/backend/commands/copy.c]: I would prefer a more descriptive log message than just `Command: %s` here, to make the logline more selfcontained, for example `COPY TO/FROM program command: %s`

File: src/backend/access/external/url_execute.c [hunk 8]
```diff
@@ -278,6 +292,7 @@ url_execute_fopen(char *url, bool forwrite, extvar_t *ev, CopyState pstate)
 	/* Restore process interval timers */
 	restoreTimers(&savetimers);
 
+	elog(DEBUG5, "Command: %s", file->shexec);
```
Comment (danielgustafsson) [inline src/backend/access/external/url_execute.c]: I would prefer a more descriptive log message than just `Command: %s` here, to make the logline more selfcontained, for example `External table command: %s`

File: src/test/regress/input/external_table.source [hunk 9]
```diff
@@ -108,8 +110,21 @@ select * from check_env;
 
 drop external table check_ps;
 drop external table check_env;
+drop external table check_qry;
```
Comment (danielgustafsson) [inline src/test/regress/input/external_table.source]: This table doesn't exist AFAICT, but since it's in an ignore block this fails to error out. Let's remove this though.

File: src/test/regress/input/external_table.source [hunk 10]
```diff
@@ -108,8 +110,21 @@ select * from check_env;
...
 
 -- end_ignore
+CREATE EXTERNAL WEB TABLE table_env (val TEXT)
+  EXECUTE E'env' ON SEGMENT 0
+  FORMAT 'TEXT' (ESCAPE 'OFF');
+SELECT * FROM table_env WHERE val LIKE 'GP_QUERY%' ORDER BY val ASC;
+SELECT * FROM table_env WHERE val LIKE 'GP_QUERY%\%' ESCAPE '&' ORDER BY val ASC;
+
+-- echo will behaves differently on different platforms, force to use bash with -E option
```
Comment (danielgustafsson) [inline src/test/regress/input/external_table.source]: s/behaves/behave/, ie: "will behave differently"

File: src/test/regress/input/external_table.source [hunk 11]
```diff
@@ -108,8 +110,21 @@ select * from check_env;
...
+CREATE EXTERNAL WEB TABLE table_qry (val TEXT)
+  EXECUTE E'/bin/bash -c ''echo -E "$GP_QUERY_STRING"''' ON SEGMENT 0
```
Comment (danielgustafsson) [inline src/test/regress/input/external_table.source]: We don't know if bash is indeed at `/bin/bash`, and have worked hard to remove such assumptions since it greatly reduce our portability for no gain. See for example 3ae06a9b6dce61128632b8d6a974d8a166afe764. Can you use a more portable approach, such as `/usr/bin/env bash -c .. ` instead?

File: src/test/regress/input/gpcopy.source [hunk 12]
```diff
@@ -1250,3 +1250,23 @@ copy broken_type_test to '/tmp/g';
...
+-- Perform these tests in a funnily named database, to test
+-- escaping
+set client_min_messages='warning';
+DROP DATABASE IF EXISTS "funny copy""db'with\\quotes";
+reset client_min_messages;
+CREATE DATABASE "funny copy""db'with\\quotes";
+
+\c "funny copy""db'with\\quotes"
+
+COPY (SELECT 'data1') TO PROGRAM 'cat > /tmp/gpcopyenvtest; /bin/bash -c ''echo -E database in COPY TO: $GP_DATABASE >> /tmp/gpcopyenvtest '' ' ESCAPE 'OFF';
```
Comment (danielgustafsson) [inline src/test/regress/input/gpcopy.source]: While it's not a practice introduced in this patch, I really dislike us polluting the global `/tmp` with testdata. It's not a parallel safe behaviour since two ICW's running at the same time risk trampling on each other. We should really be using a per-installation temporary directory for this, but that will perhaps be for a separate patch to clean up.

File: src/test/regress/output/gpcopy.source (hunk 4 above)
Comment (liming01) [reply src/test/regress/output/gpcopy.source]: Added another test case using printf. Thanks.
